resolver/resolver.go

```diff
@@ -354,111 +347,135 @@ func (res *Resolver) HandleNonMesos(w dns.ResponseWriter, r *dns.Msg) {
 // question with resource answer(s)
 // it can handle {A, SRV, ANY}
 func (res *Resolver) HandleMesos(w dns.ResponseWriter, r *dns.Msg) {
-	var err error
-
-	dom := strings.ToLower(cleanWild(r.Question[0].Name))
-	qType := r.Question[0].Qtype
+	logging.CurLog.MesosRequests.Inc()
 
-	m := new(dns.Msg)
-	m.Authoritative = true
-	m.RecursionAvailable = res.config.RecurseOn
+	m := &dns.Msg{MsgHdr: dns.MsgHdr{
+		Authoritative:      true,
+		RecursionAvailable: res.config.RecurseOn,
+	}}
 	m.SetReply(r)
 
+	var err error
 	rs := res.records()
+	name := strings.ToLower(cleanWild(r.Question[0].Name))
+	switch r.Question[0].Qtype {
+	case dns.TypeSRV:
+		err = res.handleSRV(rs, name, m, r)
+	case dns.TypeA:
+		err = res.handleA(rs, name, m)
+	case dns.TypeSOA:
+		err = res.handleSOA(m, r)
+	case dns.TypeNS:
+		err = res.handleNS(m, r)
+	case dns.TypeANY:
+		errs := multiError{
+			res.handleSRV(rs, name, m, r),
+			res.handleA(rs, name, m),
+			res.handleSOA(m, r),
+			res.handleNS(m, r),
+		}
+		if !errs.Empty() {
+			err = errs
+		}
+	}
 
-	// SRV requests
-	if (qType == dns.TypeSRV) || (qType == dns.TypeANY) {
-		for _, srv := range rs.SRVs[dom] {
-			rr, err := res.formatSRV(r.Question[0].Name, srv)
-			if err != nil {
-				logging.Error.Println(err)
-			} else {
-				m.Answer = append(m.Answer, rr)
-				// return one corresponding A record add additional info
-				host := strings.Split(srv, ":")[0]
-				if len(rs.As[host]) != 0 {
-					rr, err := res.formatA(host, rs.As[host][0])
-					if err != nil {
-						logging.Error.Println(err)
-					} else {
-						m.Extra = append(m.Extra, rr)
-					}
-				}
+	if len(m.Answer) == 0 {
+		err = res.handleEmpty(rs, name, m, r)
+	} else {
+		shuffleAnswers(m.Answer)
+		logging.CurLog.MesosSuccess.Inc()
+	}
 
-			}
-		}
+	if err != nil {
```

Comment:
Shouldn't the error case take precedence over the len==0 case? Now a handler which throws an error will possibly also return an empty answer and then the error never shows up.